Crafted COM_STMT_EXECUTE can trigger panic in `server.parseBinaryParams`

## Bug Report

### 1. Minimal reproduce step (Required)

1. Start a TiDB server.
2. Install the client dependency:
```bash
pip install pymysql
```
3. Save and run the attached PoC script:
```bash
python3 tidb_panic_poc.py --host <tidb_host> --port 4000 --user root
```
4. The script will:
   - create database `poc_db`
   - create table `t1`
   - run:
    ```sql
   CREATE TABLE t1 (
     c1 int,
     c2 text,
     PRIMARY KEY(c1),
     UNIQUE(c1)
   );
   ALTER TABLE t1 ADD COLUMN c3 int;
   ALTER TABLE t1 ADD COLUMN c4 int;
    ```
   - send one `COM_STMT_PREPARE`
   - send two crafted `COM_STMT_EXECUTE` packets over the MySQL protocol
5. On my side, the panic is triggered by the second `COM_STMT_EXECUTE`.

[tidb_panic_poc.py](https://github.com/user-attachments/files/26396798/tidb_panic_poc.py)

PoC summary:

- The prepared statement contains 8 bound parameters.
- `COM_STMT_EXECUTE #1` uses `new_params_bound_flag=1` with one NULL bitmap.
- `COM_STMT_EXECUTE #2` uses `new_params_bound_flag=0` with a different NULL bitmap and reuses cached parameter types.
- This causes a server-side panic in protocol-layer parameter parsing.


### 2. What did you expect to see? (Required)

Expect no crashes.

### 3. What did you see instead (Required)

TiDB panicked while handling `COM_STMT_EXECUTE`.

The server log shows:
```
[2026/03/27 03:34:22.440 +00:00] [ERROR] [conn.go:1042] ["connection running loop panic"] 
[conn=2092958294] [session_alias=] 
[err="runtime error: slice bounds out of range [:-3475882810334479267]"] 
[stack="github.com/pingcap/tidb/pkg/server.(*clientConn).Run.func1
	/data/db/vendor/tidb/tidb/pkg/server/conn.go:1045
runtime.gopanic
	/home/fzx/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/runtime/panic.go:785
runtime.goPanicSliceAcap
	/home/fzx/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/src/runtime/panic.go:141
github.com/pingcap/tidb/pkg/server.parseBinaryParams
	/data/db/vendor/tidb/tidb/pkg/server/conn_stmt_params.go:139
github.com/pingcap/tidb/pkg/server.(*clientConn).handleStmtExecute
	/data/db/vendor/tidb/tidb/pkg/server/conn_stmt.go:209
github.com/pingcap/tidb/pkg/server.(*clientConn).dispatch
	/data/db/vendor/tidb/tidb/pkg/server/conn.go:1406
github.com/pingcap/tidb/pkg/server.(*clientConn).Run
	/data/db/vendor/tidb/tidb/pkg/server/conn.go:1148
github.com/pingcap/tidb/pkg/server.(*Server).onConn
	/data/db/vendor/tidb/tidb/pkg/server/server.go:741"]
```
### 4. What is your TiDB version? (Required)

Please replace this with the exact output of `SELECT tidb_version();` from the affected server.
```
Release Version: v8.5.3
Edition: Community
Git Commit Hash: dc2548aac79a712265e831cff2a3a896bc0a5a38
Git Branch: HEAD
UTC Build Time: 2025-09-26 10:07:20
GoVersion: go1.23.6
Race Enabled: false
Check Table Before Drop: false
Store: tikv
```

We are the BASS team from the School of Cyber Science and Technology at Beihang University. Our main focus is on system software security, operating systems, and program analysis research, as well as the development of automated program testing frameworks for detecting software defects. Using our self-developed database vulnerability testing tool, we have identified the above-mentioned vulnerabilities in TiDB that may lead to database crashes.

--- a/pkg/server/conn_stmt_params.go
+++ b/pkg/server/conn_stmt_params.go
@@ -26,6 +26,35 @@ import (
 
 var errUnknownFieldType = dbterror.ClassServer.NewStd(errno.ErrUnknownFieldType)
 
+func parseLengthEncodedParamValueHeader(paramValues []byte, pos int) (length uint64, isNull bool, nextPos int, err error) {
+	if len(paramValues) <= pos {
+		return 0, false, pos, mysql.ErrMalformPacket
+	}
+	need := 1
+	switch paramValues[pos] {
+	case 0xfc:
+		need = 3
+	case 0xfd:
+		need = 4
+	case 0xfe:
+		need = 9
+	}
+	if len(paramValues)-pos < need {
+		return 0, false, pos, mysql.ErrMalformPacket
+	}
+	var n int
+	length, isNull, n = util2.ParseLengthEncodedInt(paramValues[pos:])
+	return length, isNull, pos + n, nil
+}
+
+func takeBinaryParamValue(paramValues []byte, pos int, length uint64) ([]byte, int, error) {
+	if pos > len(paramValues) || length > uint64(len(paramValues)-pos) {
+		return nil, pos, mysql.ErrMalformPacket
+	}
+	end := pos + int(length)
+	return paramValues[pos:end], end, nil
+}
+
 // parseBinaryParams decodes the binary params according to the protocol
 func parseBinaryParams(params []param.BinaryParam, boundParams [][]byte, nullBitmap, paramTypes, paramValues []byte, enc *util2.InputDecoder) (err error) {
 	pos := 0
@@ -106,42 +135,36 @@ func parseBinaryParams(params []param.BinaryParam, boundParams [][]byte, nullBit
 			length = uint64(paramValues[pos])
 			pos++
 		case mysql.TypeNewDecimal, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-			if len(paramValues) < (pos + 1) {
-				err = mysql.ErrMalformPacket
+			length, isNull, pos, err = parseLengthEncodedParamValueHeader(paramValues, pos)
+			if err != nil {
 				return
 			}
-			var n int
-			length, isNull, n = util2.ParseLengthEncodedInt(paramValues[pos:])
-			pos += n
 		case mysql.TypeUnspecified, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString,
 			mysql.TypeEnum, mysql.TypeSet, mysql.TypeGeometry, mysql.TypeBit:
-			if len(paramValues) < (pos + 1) {
-				err = mysql.ErrMalformPacket
+			length, isNull, pos, err = parseLengthEncodedParamValueHeader(paramValues, pos)
+			if err != nil {
 				return
 			}
-			var n int
-			length, isNull, n = util2.ParseLengthEncodedInt(paramValues[pos:])
-			pos += n
 			decodeWithDecoder = true
 		default:
 			err = errUnknownFieldType.GenWithStack("stmt unknown field type %d", tp)
 			return
 		}
 
-		if len(paramValues) < (pos + int(length)) {
-			err = mysql.ErrMalformPacket
-			return
+		val, nextPos, err := takeBinaryParamValue(paramValues, pos, length)
+		if err != nil {
+			return err
 		}
 		params[i] = param.BinaryParam{
 			Tp:         tp,
 			IsUnsigned: isUnsigned,
 			IsNull:     isNull,
-			Val:        paramValues[pos : pos+int(length)],
+			Val:        val,
 		}
 		if decodeWithDecoder {
 			params[i].Val = enc.DecodeInput(params[i].Val)
 		}
-		pos += int(length)
+		pos = nextPos
 	}
 	return
 }

--- a/pkg/server/conn_stmt_params.go
+++ b/pkg/server/conn_stmt_params.go
@@ -26,27 +26,6 @@ import (
 
 var errUnknownFieldType = dbterror.ClassServer.NewStd(errno.ErrUnknownFieldType)
 
-func parseLengthEncodedParamValueHeader(paramValues []byte, pos int) (length uint64, isNull bool, nextPos int, err error) {
-	if len(paramValues) <= pos {
-		return 0, false, pos, mysql.ErrMalformPacket
-	}
-	need := 1
-	switch paramValues[pos] {
-	case 0xfc:
-		need = 3
-	case 0xfd:
-		need = 4
-	case 0xfe:
-		need = 9
-	}
-	if len(paramValues)-pos < need {
-		return 0, false, pos, mysql.ErrMalformPacket
-	}
-	var n int
-	length, isNull, n = util2.ParseLengthEncodedInt(paramValues[pos:])
-	return length, isNull, pos + n, nil
-}
-
 func takeBinaryParamValue(paramValues []byte, pos int, length uint64) ([]byte, int, error) {
 	if pos > len(paramValues) || length > uint64(len(paramValues)-pos) {
 		return nil, pos, mysql.ErrMalformPacket
@@ -135,16 +114,20 @@ func parseBinaryParams(params []param.BinaryParam, boundParams [][]byte, nullBit
 			length = uint64(paramValues[pos])
 			pos++
 		case mysql.TypeNewDecimal, mysql.TypeBlob, mysql.TypeTinyBlob, mysql.TypeMediumBlob, mysql.TypeLongBlob:
-			length, isNull, pos, err = parseLengthEncodedParamValueHeader(paramValues, pos)
+			var n int
+			length, isNull, n, err = util2.ParseLengthEncodedInt(paramValues[pos:])
 			if err != nil {
-				return
+				return mysql.ErrMalformPacket
 			}
+			pos += n
 		case mysql.TypeUnspecified, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeString,
 			mysql.TypeEnum, mysql.TypeSet, mysql.TypeGeometry, mysql.TypeBit:
-			length, isNull, pos, err = parseLengthEncodedParamValueHeader(paramValues, pos)
+			var n int
+			length, isNull, n, err = util2.ParseLengthEncodedInt(paramValues[pos:])
 			if err != nil {
-				return
+				return mysql.ErrMalformPacket
 			}
+			pos += n
 			decodeWithDecoder = true
 		default:
 			err = errUnknownFieldType.GenWithStack("stmt unknown field type %d", tp)

--- a/pkg/server/conn_stmt_params_test.go
+++ b/pkg/server/conn_stmt_params_test.go
@@ -282,6 +282,40 @@ func TestParseExecArgs(t *testing.T) {
 	}
 }
 
+func TestParseExecArgsMalformedLengthEncodedParam(t *testing.T) {
+	tests := []struct {
+		name        string
+		paramValues []byte
+	}{
+		{
+			name:        "truncated length-encoded uint64 header",
+			paramValues: []byte{0xfe},
+		},
+		{
+			name: "overflowing length-encoded uint64",
+			// length = 1<<63, which previously could slip past `pos + int(length)` checks and panic on slicing.
+			paramValues: []byte{0xfe, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require.NotPanics(t, func() {
+				args := expression.Args2Expressions4Test(1)
+				err := decodeAndParse(
+					types.DefaultStmtNoWarningContext,
+					args,
+					[][]byte{nil},
+					[]byte{0x0},
+					[]byte{mysql.TypeVarString, 0},
+					tt.paramValues,
+					nil,
+				)
+				require.Truef(t, terror.ErrorEqual(err, mysql.ErrMalformPacket), "err %v", err)
+			})
+		})
+	}
+}
+
 func TestParseExecArgsAndEncode(t *testing.T) {
 	dt := expression.Args2Expressions4Test(1)
 	err := decodeAndParse(types.DefaultStmtNoWarningContext,

--- a/pkg/server/conn_test.go
+++ b/pkg/server/conn_test.go
@@ -547,6 +547,37 @@ func TestHandshakeResponseCompatibilityAndFailurePaths(t *testing.T) {
 		require.ErrorIs(t, err, mysql.ErrMalformPacket)
 	})
 
+	t.Run("truncated connect attrs length header", func(t *testing.T) {
+		data := buildHandshakeResponsePacket(mysql.ClientProtocol41|mysql.ClientConnectAtts, nil, nil)
+		data[len(data)-1] = 0xfe
+
+		var p handshake.Response41
+		offset, err := parse.HandshakeResponseHeader(context.Background(), &p, data)
+		require.NoError(t, err)
+
+		err = parse.HandshakeResponseBody(context.Background(), &p, data, offset)
+		require.ErrorIs(t, err, mysql.ErrMalformPacket)
+	})
+
+	t.Run("truncated auth length header", func(t *testing.T) {
+		var buf bytes.Buffer
+		binary.Write(&buf, binary.LittleEndian, uint32(mysql.ClientProtocol41|mysql.ClientPluginAuthLenencClientData))
+		binary.Write(&buf, binary.LittleEndian, uint32(0)) // max packet size
+		buf.WriteByte(0)                                   // collation
+		buf.Write(make([]byte, 23))                        // reserved
+		buf.WriteString("root")
+		buf.WriteByte(0)    // user null-terminated
+		buf.WriteByte(0xfe) // truncated auth length-encoded header
+		data := buf.Bytes()
+
+		var p handshake.Response41
+		offset, err := parse.HandshakeResponseHeader(context.Background(), &p, data)
+		require.NoError(t, err)
+
+		err = parse.HandshakeResponseBody(context.Background(), &p, data, offset)
+		require.ErrorIs(t, err, mysql.ErrMalformPacket)
+	})
+
 	t.Run("oversize connect attrs declaration rejected", func(t *testing.T) {
 		declaredLen := uint64(1<<20 + 1)
 		data := buildHandshakeResponsePacket(mysql.ClientProtocol41|mysql.ClientConnectAtts, nil, &declaredLen)

--- a/pkg/server/conn_test.go
+++ b/pkg/server/conn_test.go
@@ -561,7 +561,7 @@ func TestHandshakeResponseCompatibilityAndFailurePaths(t *testing.T) {
 
 	t.Run("truncated auth length header", func(t *testing.T) {
 		var buf bytes.Buffer
-		binary.Write(&buf, binary.LittleEndian, uint32(mysql.ClientProtocol41|mysql.ClientPluginAuthLenencClientData))
+		binary.Write(&buf, binary.LittleEndian, mysql.ClientProtocol41|mysql.ClientPluginAuthLenencClientData)
 		binary.Write(&buf, binary.LittleEndian, uint32(0)) // max packet size
 		buf.WriteByte(0)                                   // collation
 		buf.Write(make([]byte, 23))                        // reserved

--- a/pkg/server/internal/parse/parse.go
+++ b/pkg/server/internal/parse/parse.go
@@ -92,7 +92,10 @@ func HandshakeResponseBody(ctx context.Context, packet *handshake.Response41, da
 		if data[offset] == 0x1 { // No auth data
 			offset += 2
 		} else {
-			num, null, off := util2.ParseLengthEncodedInt(data[offset:])
+			num, null, off, err := util2.ParseLengthEncodedInt(data[offset:])
+			if err != nil {
+				return mysql.ErrMalformPacket
+			}
 			offset += off
 			if !null {
 				packet.Auth = data[offset : offset+int(num)]
@@ -133,7 +136,10 @@ func HandshakeResponseBody(ctx context.Context, packet *handshake.Response41, da
 			// Defend some ill-formated packet, connection attribute is not important and can be ignored.
 			return nil
 		}
-		num, null, intOff := util2.ParseLengthEncodedInt(data[offset:])
+		num, null, intOff, err := util2.ParseLengthEncodedInt(data[offset:])
+		if err != nil {
+			return mysql.ErrMalformPacket
+		}
 		offset += intOff // Length of variable length encoded integer itself in bytes
 		if !null {
 			if num > 1<<20 { // 1 MiB hard limit

--- a/pkg/server/internal/util/util.go
+++ b/pkg/server/internal/util/util.go
@@ -66,8 +66,6 @@ func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int, err error)
 	switch b[0] {
 	// 251: NULL
 	case 0xfb:
-		n = 1
-		isNull = true
 		return 0, true, 1, nil
 
 	// 252: value of following 2
@@ -76,7 +74,6 @@ func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int, err error)
 			return 0, false, 0, io.EOF
 		}
 		num = uint64(b[1]) | uint64(b[2])<<8
-		n = 3
 		return num, false, 3, nil
 
 	// 253: value of following 3
@@ -85,7 +82,6 @@ func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int, err error)
 			return 0, false, 0, io.EOF
 		}
 		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16
-		n = 4
 		return num, false, 4, nil
 
 	// 254: value of following 8
@@ -96,7 +92,6 @@ func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int, err error)
 		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16 |
 			uint64(b[4])<<24 | uint64(b[5])<<32 | uint64(b[6])<<40 |
 			uint64(b[7])<<48 | uint64(b[8])<<56
-		n = 9
 		return num, false, 9, nil
 	}
 

--- a/pkg/server/internal/util/util.go
+++ b/pkg/server/internal/util/util.go
@@ -57,48 +57,63 @@ func ParseNullTermString(b []byte) (str []byte, remain []byte) {
 }
 
 // ParseLengthEncodedInt parses a length encoded integer.
-func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int) {
+// It returns io.EOF when the length-encoded header is truncated.
+func ParseLengthEncodedInt(b []byte) (num uint64, isNull bool, n int, err error) {
+	if len(b) == 0 {
+		return 0, false, 0, io.EOF
+	}
+
 	switch b[0] {
 	// 251: NULL
 	case 0xfb:
 		n = 1
 		isNull = true
-		return
+		return 0, true, 1, nil
 
 	// 252: value of following 2
 	case 0xfc:
+		if len(b) < 3 {
+			return 0, false, 0, io.EOF
+		}
 		num = uint64(b[1]) | uint64(b[2])<<8
 		n = 3
-		return
+		return num, false, 3, nil
 
 	// 253: value of following 3
 	case 0xfd:
+		if len(b) < 4 {
+			return 0, false, 0, io.EOF
+		}
 		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16
 		n = 4
-		return
+		return num, false, 4, nil
 
 	// 254: value of following 8
 	case 0xfe:
+		if len(b) < 9 {
+			return 0, false, 0, io.EOF
+		}
 		num = uint64(b[1]) | uint64(b[2])<<8 | uint64(b[3])<<16 |
 			uint64(b[4])<<24 | uint64(b[5])<<32 | uint64(b[6])<<40 |
 			uint64(b[7])<<48 | uint64(b[8])<<56
 		n = 9
-		return
+		return num, false, 9, nil
 	}
 
-	// https://dev.mysql.com/doc/internals/en/integer.html#length-encoded-integer: If the first byte of a packet is a length-encoded integer and its byte value is 0xfe, you must check the length of the packet to verify that it has enough space for a 8-byte integer.
-	// TODO: 0xff is undefined
+	// https://dev.mysql.com/doc/internals/en/integer.html#length-encoded-integer
+	// TODO: 0xff is undefined.
 
 	// 0-250: value of first byte
-	num = uint64(b[0])
-	n = 1
-	return
+	return uint64(b[0]), false, 1, nil
 }
 
 // ParseLengthEncodedBytes parses a length encoded byte slice.
 func ParseLengthEncodedBytes(b []byte) ([]byte, bool, int, error) {
 	// Get length
-	num, isNull, n := ParseLengthEncodedInt(b)
+	num, isNull, n, err := ParseLengthEncodedInt(b)
+	if err != nil {
+		return nil, isNull, n, err
+	}
 	if num < 1 {
 		return nil, isNull, n, nil
 	}

--- a/pkg/server/internal/util/util_test.go
+++ b/pkg/server/internal/util/util_test.go
@@ -15,6 +15,7 @@
 package util
 
 import (
+	"io"
 	"strconv"
 	"testing"
 
@@ -209,45 +210,82 @@ func TestParseLengthEncodedInt(t *testing.T) {
 		num    uint64
 		isNull bool
 		n      int
+		err    error
 	}{
 		{
 			[]byte{'\xfb'},
 			uint64(0),
 			true,
 			1,
+			nil,
 		},
 		{
 			[]byte{'\x00'},
 			uint64(0),
 			false,
 			1,
+			nil,
 		},
 		{
 			[]byte{'\xfc', '\x01', '\x02'},
 			uint64(513),
 			false,
 			3,
+			nil,
 		},
 		{
 			[]byte{'\xfd', '\x01', '\x02', '\x03'},
 			uint64(197121),
 			false,
 			4,
+			nil,
 		},
 		{
 			[]byte{'\xfe', '\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\x07', '\x08'},
 			uint64(578437695752307201),
 			false,
 			9,
+			nil,
+		},
+		{
+			[]byte{},
+			uint64(0),
+			false,
+			0,
+			io.EOF,
+		},
+		{
+			[]byte{'\xfc', '\x01'},
+			uint64(0),
+			false,
+			0,
+			io.EOF,
+		},
+		{
+			[]byte{'\xfd', '\x01', '\x02'},
+			uint64(0),
+			false,
+			0,
+			io.EOF,
+		},
+		{
+			[]byte{'\xfe'},
+			uint64(0),
+			false,
+			0,
+			io.EOF,
 		},
 	}
 
 	for _, tc := range testCases {
-		num, isNull, n := ParseLengthEncodedInt(tc.buffer)
+		num, isNull, n, err := ParseLengthEncodedInt(tc.buffer)
 		require.Equal(t, tc.num, num)
 		require.Equal(t, tc.isNull, isNull)
 		require.Equal(t, tc.n, n)
-		require.Equal(t, tc.n, LengthEncodedIntSize(tc.num))
+		require.ErrorIs(t, err, tc.err)
+		if tc.err == nil {
+			require.Equal(t, tc.n, LengthEncodedIntSize(tc.num))
+		}
 	}
 }
 
@@ -272,6 +310,13 @@ func TestParseLengthEncodedBytes(t *testing.T) {
 	require.False(t, isNull)
 	require.Equal(t, 2, n)
 	require.Equal(t, "EOF", err.Error())
+
+	buffer = []byte{'\xfe'}
+	b, isNull, n, err = ParseLengthEncodedBytes(buffer)
+	require.Nil(t, b)
+	require.False(t, isNull)
+	require.Equal(t, 0, n)
+	require.ErrorIs(t, err, io.EOF)
 }
 
 func TestParseNullTermString(t *testing.T) {

